src/ir_Kelon.cpp

```diff
@@ -0,0 +1,486 @@
+/// @param[in] toggle Whether to toggle the power state
+void IRKelonAC::setTogglePower(const bool toggle) {
+  _.PowerToggle = toggle;
+}
+
+/// Get whether toggling power will be requested
+/// @return The power toggle state
+bool IRKelonAC::getTogglePower() const {
+  return _.PowerToggle;
+}
+
+/// Set the temperature setting.
+/// @param[in] degrees The temperature in degrees celsius.
+void IRKelonAC::setTemp(const uint8_t degrees) {
+  uint8_t temp = std::max(kKelonMinTemp, degrees);
+  temp = std::min(kKelonMaxTemp, temp);
+  _previousTemp = _.Temperature;
+  _.Temperature = temp - kKelonMinTemp;
+}
+
+/// Get the current temperature setting.
+/// @return Get current setting for temp. in degrees celsius.
+uint8_t IRKelonAC::getTemp() const {
+  return _.Temperature + kKelonMinTemp;
+}
+
+/// Set the speed of the fan.
+/// @param[in] speed 0 is auto, 1-5 is the speed
+void IRKelonAC::setFan(const uint8_t speed) {
+  uint8_t fan = std::min(speed, kKelonFanMax);
+
+  _previousFan = _.Fan;
+  // Note: Kelon fan speeds are backwards! This code maps the range 0,1:3 to 0,3:1 to save the API's user's sanity.
+  _.Fan = ((static_cast<int16_t>(fan) - 4) * -1) % 4;
+}
+
+/// Get the current fan speed setting.
+/// @return The current fan speed.
+uint8_t IRKelonAC::getFan() const {
+  return ((static_cast<int16_t>(_.Fan) - 4) * -1) % 4;;
+}
+
+/// Set the dehumidification intensity.
+/// @param[in] grade has to be in the range [-2 : +2]
+void IRKelonAC::setDryGrade(const int8_t grade) {
+  int8_t drygrade = std::max(kKelonDryGradeMin, grade);
+  drygrade = std::min(kKelonDryGradeMax, drygrade);
+
+  // Two's complement is clearly too bleeding edge for this manufacturer
+  uint8_t outval;
+  if (drygrade < 0) {
+    outval = 0b100 | (-drygrade & 0b011);
+  } else {
+    outval = drygrade & 0b011;
+  }
+  _.DehumidifierGrade = outval;
+}
+
+/// Get the current dehumidification intensity setting. In smart mode, this controls the temperature adjustment.
+/// @return The current dehumidification intensity.
+int8_t IRKelonAC::getDryGrade() const {
+  auto outval = static_cast<int8_t>(_.DehumidifierGrade & 0b011);
+  if ((_.DehumidifierGrade & 0b100) == 0b100) {
+    outval *= -1;
+  }
+  return outval;
+}
+
+/// Set the desired operation mode.
+/// @param[in] mode The desired operation mode.
+void IRKelonAC::setMode(const uint8_t mode) {
+  if (_.Mode == kKelonModeSmart || _.Mode == kKelonModeFan || _.Mode == kKelonModeDry) {
+    _.Temperature = _previousTemp;
+  }
+  if (_.SuperCoolEnabled1) {
+    // Cancel supercool
+    _.SuperCoolEnabled1 = false;
+    _.SuperCoolEnabled2 = false;
+    _.Temperature = _previousTemp;
+    _.Fan = _previousFan;
+  }
+  _previousMode = _.Mode;
+
+  switch (mode) {
+    case kKelonModeSmart:
+      setTemp(26);
+      _.SmartModeEnabled = true;
+      _.Mode = mode;
+      break;
+    case kKelonModeDry:
+    case kKelonModeFan:
+      setTemp(25);
+      _.Mode = mode;
```

Comment:
I think this line is redundant, as the next `case` clause will do it due to the FALL THRU